tun/tun_illumos.go

```diff
@@ -0,0 +1,519 @@
+//go:cgo_import_dynamic libc_putmsg putmsg "libc.so"
+//go:cgo_import_dynamic libc_getmsg getmsg "libc.so"
+
+//go:linkname f_ioctl libc_ioctl
+//go:linkname f_putmsg libc_putmsg
+//go:linkname f_getmsg libc_getmsg
+
+var (
+	f_ioctl uintptr
+	f_putmsg uintptr
+	f_getmsg uintptr
+)
+
+const (
+	/*
+	 * For use with "/dev/tun":
+	 */
+	TUNNEWPPA = 0x540001
+	TUNSETPPA = 0x540002
+
+	/*
+	 * sys/stropts.h:
+	 */
+	I_STR = 0x5308
+	I_POP = 0x5303
+	I_PUSH = 0x5302
+	I_PLINK = 0x5316
+	I_PUNLINK = 0x5317
+
+	/*
+	 * sys/sockio.h:
+	 */
+	IF_UNITSEL = 0x80047336 /* set unit number */
+
+	SIOCSLIFMUXID = 0x80786984
+	SIOCGLIFMUXID = 0xc0786983
+	SIOCGLIFINDEX = 0xc0786985
+)
+
+func sysvicall6(trap, nargs, a1, a2, a3, a4, a5, a6 uintptr) (
+    r1, r2 uintptr, err syscall.Errno)
+
+func ioctl(fd int, req uint, arg uintptr) (r int, err error) {
+	r0, _, e0 := sysvicall6(uintptr(unsafe.Pointer(&f_ioctl)), 3,
+	    uintptr(fd), uintptr(req), arg, 0, 0, 0)
+
+	r = int(r0)
+	if e0 != 0 {
+		err = e0
+	}
+
+	return
+}
+
+func putmsg(fd int, ctlptr uintptr, dataptr uintptr, flags int) (
+    r int, err error) {
+	r0, _, e0 := sysvicall6(uintptr(unsafe.Pointer(&f_putmsg)), 4,
+	    uintptr(fd), ctlptr, dataptr, uintptr(flags), 0, 0)
+
+	r = int(r0)
+	if e0 != 0 {
+		err = e0
+	}
+
+	return
+}
+
+func getmsg(fd int, ctlptr uintptr, dataptr uintptr, flagsp uintptr) (
+    r int, err error) {
+	r0, _, e0 := sysvicall6(uintptr(unsafe.Pointer(&f_getmsg)), 4,
+	    uintptr(fd), ctlptr, dataptr, flagsp, 0, 0)
+
+	r = int(r0)
+	if e0 != 0 {
+		err = e0
+	}
+
+	return
+}
+
+
+/*
+ * WIREGUARD/TUN INTERFACE
+ */
+
+type NativeTun struct {
+	tunFile *os.File	/* TUN device file */
+	ip_fd int		/* IP device fd */
+	name string		/* Interface name */
+	mtu int
+
+	events chan Event
+	errors chan error
+}
+
+// type Device interface {
+// 	File() *os.File                 // returns the file descriptor of the device
+// 	Read([]byte, int) (int, error)  // read a packet from the device (without any additional headers)
+// 	Write([]byte, int) (int, error) // writes a packet to the device (without any additional headers)
+// 	MTU() (int, error)              // returns the MTU of the device
+// 	Name() (string, error)          // fetches and returns the current name
+// 	Events() chan Event          // returns a constant channel of events related to the device
+// 	Close() error                   // stops the device and closes the event channel
+// }
+
+func (tun *NativeTun) Name() (string, error) {
+	return tun.name, nil
+}
+
+func (tun *NativeTun) Flush() error {
+        // TODO: can flushing be implemented by buffering and using sendmmsg?
+        return nil
+}
+
+func (tun *NativeTun) Read(buf []byte, offset int) (int, error) {
+	select {
+	case err := <-tun.errors:
+		return 0, err
+
+	default:
+		return tun.read_tun(buf[offset:])
+	}
+}
+
+func (tun *NativeTun) Write(buf []byte, offset int) (int, error) {
+	return tun.write_tun(buf[offset:])
```

Comment:
There's discussion of that further down, but if these helpers need to continue doing deep unsafe things, the encapsulation of that into a helper is nice in that it isolates the unsafety a little. Hopefully we can dispense with the unsafe stuff instead.

If we do end up keeping the helpers, I would put them right next to their parent functions, so that a reader can go straight from the friendly Go Read/Write's implementation into the tricky internals, if they're trying to follow the entirety of the read/write path.